### PR TITLE
Update Android build scripts

### PR DESCRIPTION
**Pull request by robertxgray**
I have updated the Android build scripts. The tests have been performed with DCSS 0.26.1, SDK for Android 11 (API level 30) and NDK version r22b.

The build process has little changes, it's a bit more simple because now the SDK includes all the required tools. Android Studio can also be used. For more details you can check the updated version of docs/develop/android.txt.

Worshippers of Xom can download the apk from here: https://github.com/robertxgray/crawl/releases

Some changes are required in other repositories:
- https://github.com/crawl/crawl-lua/pull/2
- https://github.com/crawl/crawl-sdl2-image/pull/2
- https://github.com/crawl/crawl-sdl2-mixer/pull/1

### CODE REVIEW

**Comment by rawlins** [comment]
Thanks for doing this! You should be aware that the current set of active committers have no experience whatsoever with android, so we can't really vet this, and it may go stale pretty quickly if there's not someone outside contributing updates (`TOUCH_UI` builds are also not well-cared-for). But I would vote for merging these changes nonetheless.

To help both whoever merges this and future committers can you rebase to add some details to the commit message itself explaining what is changed here? E.g. content currently in the PR message plus stuff like (caveat I am only 80% sure what gradle even is so I may be wording this wrong):
 * added a gradle build setup for the android app
 * removed previous build.xml
 * updated various #ifdefs for `__ANDROID__` and `TOUCH_UI` to handle current android builds

plus whatever else you deem appropriate? Commits that would give someone a higher-level overview about what exactly was needed to get this to work.

**Comment by robertxgray** [comment]
Sure, I just updated the commit message.

I'm a big fan of Barlow's ascii port for Android and I've been playing it for years. Now he's lacking time to maintain it so I thought I could step in. I didn't even know there was a tiles version for Android, I just found out while trying to build the code for the ascii port because I was unable to do it his way.

Now I'm not sure which version I'll be playing in the future, but both ports can benefit from these changes. I really need DCSS in my phone, otherwise it's as useless as a ring of attention.

**Comment by robertxgray** [comment]
I'm happy to announce what I think it's a big improvement over the first version of the Android port. The user experience should be far more pleasant with bigger texts, better controls and some bug fixes. Please check the commit description for more details.

This release also aims for easier maintenance, by reducing the differences between the Android build and the rest. I tested every __ANDROID__ and TOUCH_UI directive, those which are not indispensable or don't work as expected have been removed. The remains of TOUCH_UI code have been merged in the Android build because I don't think they'll be useful elsewhere. New features have been implemented in Java code when possible.

Now I'll be heading to the tavern to find some beta testers :D

**Review by rawlins — CHANGES_REQUESTED** [review]
Sounds like good progress! I've gone ahead and merged the contrib updates, because those were relatively easy, so it should be possible to adjust the changes here to not touch the contrib directory, which will make eventual merge easier. Possibly from the feedback in the reddit post on this, possibly it isn't quite time to merge the main PR regardless of feedback below?

I did a code review and to the best of my knowledge this looks mostly fine. I think it is probably ok to remove `TOUCH_UI`; in principle this is used by non-android devs to sanity check UI stuff for small screens, but I'm not sure how often that really happens in practice. The changes involved in this part didn't look too major to me anyways. There's one comment where I spotted what looks like a mistake. Here's a few further notes:

* `gradle-wrapper.jar`: I see from googling that this is intended to be committed under version control as (if I'm getting it right) a way of pinning dependencies. However, this is controversial for the exact same reasons that led to me googling this (see e.g. https://stackoverflow.com/questions/20348451/why-should-the-gradle-wrapper-be-committed-to-vcs) and I'm not convinced that committing it to this repository is the right thing to do. In general we don't commit generated files if at all possible, especially binaries. In addition, this seems to me to present some licensing issues (that have come up before in this project); distributing this file in a GPL2 project obligates us to provide the source for this somehow in human-readable form. (In principle this may or may not be technically satisfied by just providing a link but in practice this will be likely to generate complaints from open-source advocates and has in the past.) Given these two issues, is it at all possible just to provide an automated or manual way of downloading the file, and describe that? (If it were to be included, it would need to be excluded from the `checkwhite` script and probably also from the nodeps tarbell generation in `Makefile`. Also, it would need some indication of its licensing, which I believe is also ASL2.)

* License issues: aside from the above, I highlighted a few specific places where I think there are issues but any code coming from another project should have its licensing status made clear, in case I did miss something. I'm more ok with including generated code that takes the form of human-readable scripts.

* I ran CI for this PR and it looks mostly fine but there's quite a few `checkwhite` errors. (See https://github.com/crawl/crawl/blob/master/crawl-ref/docs/develop/coding_conventions.md, but once you fix the jar issue one way or another you can probably just run this script with `-a` and automate fixing everything else there.)

File: crawl-ref/source/android-project/gradlew
Comment (robertxgray) [reply crawl-ref/source/android-project/gradlew:1]: I confirmed Android Studio doesn't need these files. We could remove support for the automated build process and depend on Android Studio. The non-signed APKs generated by the Makefile aren't useful anyway. Would this commit solve the licensing issues?
https://github.com/robertxgray/crawl/commit/77fc2a0309cacdaf7e6e263135651e8bffa54e41

Thank you for your review. I'll look into the other issues ASAP.

File: crawl-ref/source/ability.cc
Comment (robertxgray) [reply crawl-ref/source/ability.cc]: Oops :(
Thank you, it should be fixed in the next commit.

File: crawl-ref/source/android-project/gradlew
Comment (robertxgray) [reply crawl-ref/source/android-project/gradlew]: The removal of the script is included in the new commit.

File: crawl-ref/source/windowmanager-sdl.cc
Comment (robertxgray) [reply crawl-ref/source/windowmanager-sdl.cc:447]: The removal of the script is included in the new commit.

File: crawl-ref/source/android-project/gradlew.bat
Comment (robertxgray) [reply crawl-ref/source/android-project/gradlew.bat]: The removal of the script is included in the new commit.

File: crawl-ref/source/windowmanager-sdl.cc
Comment (robertxgray) [reply crawl-ref/source/windowmanager-sdl.cc:447]: Sorry, I abused copy-paste. I'm proposing a new configuration option `tile_window_minimum` that disables the window size restrictions when set to false. This allows testing without the need to recompile the code.

Example:
```
tile_window_width = 400
tile_window_height = 400
tile_window_minimum = false
```

Screenshot

**Comment by robertxgray** [comment]
I hope I'm not making a mess. I pushed another commit with solution proposals for the issues you found and then merged the main branch to solve the conflicts. The changes are detailled in the individual conversations.

**Comment by robertxgray** [comment]
I ignored the tile_use_small_layout so far, because it was broken and I didn't know if it was any good. But now I have been working on it and it will be ready for the next beta version.

**Comment by robertxgray** [comment]
The next release is ready! All the problems detected during the first beta should be fixed and it also features a reworked small layout and custom keyboard. I'll make the announce when the apk is ready, it takes very long to build though.

All the Android related commits have been squashed.

Screenshot_2021-05-23-17-21-44

**Comment by robertxgray** [comment]
I'm happy with the current status of the port and the user feedback is very positive so far. My plan is to start a feature freeze so we can focus on the merge. I am also learning about how to publish the game on the play store.

**Comment by Goratrix** [comment]
Just wanted to say THANK YOU for this work! The latest build is very nicely playable, the small layout is a huge improvement in playability and UX on the mobile screen.

**Comment by rawlins** [comment]
> I'm happy with the current status of the port and the user feedback is very positive so far. My plan is to start a feature freeze so we can focus on the merge. I am also learning about how to publish the game on the play store.

I haven't had much time to look at this, sorry, but this looks great and it sounds like it is probably close to a merge.

One further thought I had, which maybe is out of scope for this immediate PR (I'm not sure what would be involved at all): is there a way to do this build in CI (that is, in Github Actions)?

**Comment by robertxgray** [comment]
I don't know Github Actions and my knowledge about CI is very basic, but I think the same runner used for the tiles version of Linux should be able to create the Android project with: "ANDROID=1 android" parameters. I'm using Fedora 33 without any special setup.

For a complete build we just need to run an aditional step with Gradle, but it depends on the SDK and NDK. Note that building the game for every Android ABI takes over one hour with my (slightly obsolete) i5-4670 CPU and uses almost 20GB of storage, including the SDK and NDK.

**Comment by robertxgray** [comment]
Just pushed a small fix to solve a bug in the small layout. The game crashes when wielding a launcher while quivering something
different than the ammo.

**Comment by robertxgray** [comment]
It's now available on Google Play: https://play.google.com/store/apps/details?id=org.develz.crawl

**Comment by robertxgray** [comment]
The Android port cannot read the bone files on dungeon generation, this is because fopen is used instead of fopen_u. I have also replaced other fopen calls that I found, most of them on wizard, test and debug functions. Maybe the Windows port also benefits from this fix?

Finally, I also set the external storage path as the working directory on Android. This fixes the wizard fighting simulator and probably other problems.

**Comment by robertxgray** [comment]
> Note that building the game for every Android ABI takes over one hour...

The default configuration uses a single core for building with NDK, it can be improved if more cores are available :P
Now it takes 15min in my workstation.

**Comment by robertxgray** [comment]
The UI looks too small on landscape mode and even crashes on some devices when the virtual keyboard is shown. Now the keyboard shows only in portrait mode. Conflicts solved and commits squashed again.

**Comment by robertxgray** [comment]
I had some extra conflict problems because 0.26.1 is still used as the base for the Android release, but I was finally able to pass all the checks.

**Comment by robertxgray** [comment]
Last push was a tiny change in the Makefile so the number of jobs used by the NDK is obtained with nproc.

**Comment by robertxgray** [comment]
I have pushed a tiny fix for a small but annoying bug in the tabbed interface. I keep squashing the commits because that's what I've been asked to do in other projects, but never with such big commits. Please tell me if you prefer a different workflow or if there's anything else I can do to help with the merge. I am not aware of any pending tasks.

I made the pull request also available on the Bloatcrawl 2 repository for fun.
https://github.com/Hellmonk/bloatcrawl2/pull/63
https://play.google.com/store/apps/details?id=org.develz.bloatcrawl2

**Comment by robertxgray** [comment]
The Android version has been updated to 0.27. I included some other minor changes and bugfixes, described in the commit comment. The small layout shouldn't crash anymore on landscape mode, so I enabled the keyboard again, there was at least one person missing it.

**Comment by robertxgray** [comment]
I realized I was using the wrong .gitignore file while working on #2049.

File: crawl-ref/source/initfile.cc
Comment (linonetwo) [inline crawl-ref/source/initfile.cc:4029]: Hi, @robertxgray , I'm working for CDDA and trying to upgrade its android build to adapt Android 11 https://github.com/CleverRaven/Cataclysm-DDA/issues/49310#L528

Could you please teach me how to make game save to the /media folder? AFAIK, this SDL_AndroidGetExternalStoragePath will return something like `/sdcard/Android/data/org.develz.crawl/saves/`, while Android 11 require that saves should be inside `/media/org.develz.crawl/saves/` so player can backup/modify/sync their save files easily.

Comment (robertxgray) [reply crawl-ref/source/initfile.cc:4029]: Hi again!

This is what I found in the documentation.

The DCSS app includes the WRITE_EXTERNAL_STORAGE permission in the manifest, but [it's no longer needed](https://developer.android.com/guide/topics/manifest/uses-permission-element) for the app to write it's own application-specific directories since Android 4.4.

This permission can be used to write in other directories, but [it must be explicitly granted by the user](https://developer.android.com/training/permissions/requesting) on devices running Android 6.0 or higher. I made some successful tests, for example using /storage/emulated/0/dcss instead of the path returned by SDL_AndroidGetExternalStoragePath.

The application-specific directories cannot be browsed on Android 11 using the stock file manager, but it can be done with other file managers like [Speed Software's Explorer](https://play.google.com/store/apps/details?id=com.speedsoftware.explorer).

I didn't find anything about /media/org.develz.crawl/saves, unless you mean the [MediaStore API](https://developer.android.com/training/data-storage/shared/media) which can be used to write on media paths without extra permissions. I'm not sure if it's a viable solution because it's designed for media files and I couldn't find an API for the native interface.

I think the best solution would be using the app-specific directory by default but giving the users an option to use another path instead. Only if they choose another path they'll need to grant extra permissions to the app. I prefer not requesting extra permissions unless strictly necessary.

But I don't plan to include this feature in DCSS for now, I don't even know where to include the option because the game doesn't have any configuration menus. I hope not many users want to access the files while using Android 11 and they can still do it by installing another file manager. What I'm going to do, is restricting the WRITE_EXTERNAL_STORAGE permission to Android 4.3 or lower in the next update.

Kind regards.

**Comment by robertxgray** [comment]
Conflicts are arising again, I'm waiting for some more to group the review tasks, but just tell me if you need them resolved.

**Comment by robertxgray** [comment]
Hide tab icons by default demo.

https://user-images.githubusercontent.com/9676118/145578349-42906efc-61d7-49b4-8e8b-fbc68b0d6421.mp4

**Comment by robertxgray** [comment]
I learned a little bit about GitHub Actions and it turns out the ubuntu runner includes everything we need for the Android build, it was trivial to set up.

I configured a simple debug build, not very useful for players but it may prevent the Android port from breaking. After the merge, we can consider automatic releases in the Play Store and maybe even another Android app based on trunk.

**Comment by robertxgray** [comment]
Responsive layout demo.

https://user-images.githubusercontent.com/9676118/187243965-f12b022d-dd34-4c27-a3b3-69beddbe0f0c.mp4

**Comment by rawlins** [comment]
We should finally merge this -- what will that take in terms of conflict resolution? I see the branch is relatively up to date so hopefully not too much.

**Comment by robertxgray** [comment]
The Android port is currently using 0.29.1 and there are no conflicts with trunk. I just made a local merge and performed basic tests on Android and Linux without issues.

I haven't been following the development on trunk since the last release. Can you think of any changes or new features where I should focus?

**Comment by rawlins** [comment]
Re changes for 0.30, nothing that would majorly affect android is coming to mind, but I may be forgetting something (most of my effort has been webtiles lately). There is a new options editing PR that just dropped.

I'm still seeing conflicts in the github UI -- I haven't tried to do a merge locally yet though, it may not be anything major; I can take a look.

**Comment by rawlins** [comment]
I have merged this -- thanks again for your work on the android version, @robertxgray !

**Pull request closed without merging**
